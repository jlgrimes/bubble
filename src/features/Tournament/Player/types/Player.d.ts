import { PlayerMatch } from "./PlayerMatch";
import { Record } from "./Record";

export interface Player {
  /**
   * Unique identifier of the player. Bye player has ID 'bye'
   */
  id: string;
  /**
   * Name of the player. Ex. Kenny
   */
  name: string;
  /**
   * Matches the player has played. Starts as empty array
   */
  matches: PlayerMatch[];
  /**
   * The record a player has. Should be updated alongside matches.
   */
  record: Record;
  /**
   * Number of match points a player has
   */
  matchPoints: number;
  /**
   * Dropped status. If true, player is dropped, do not include in pairings.
   */
  dropped?: boolean;
}