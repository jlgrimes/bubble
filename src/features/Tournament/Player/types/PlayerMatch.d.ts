import { MatchResult } from "../../Pairings/types/MatchResult"

/**
 * PlayerMatch is the match object stored in the player objects when the round completes.
 * Match object is mapped down to the PlayerMatch object, accounting for wins/ties/etc.
 */
 export interface PlayerMatch {
  /**
   * Id of opponent
   */
  opponentId: String,
  /**
   * Result of the match. Current player against the opponentId player.
   */
  result: MatchResult
}