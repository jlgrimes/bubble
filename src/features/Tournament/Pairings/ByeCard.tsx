import Typography from "@mui/material/Typography";
import { PairingHeaderCard } from "./PairingHeaderCard";

export const ByeCard = () => {
  return (
    <PairingHeaderCard aria-label='Bye'>
      <Typography>Bye</Typography>
    </PairingHeaderCard>
  );
};