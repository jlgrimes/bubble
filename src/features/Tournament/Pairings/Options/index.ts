import { PairingFilters } from './PairingFilters';
import { PairingSearch } from './PairingSearch';

export { PairingFilters, PairingSearch };
