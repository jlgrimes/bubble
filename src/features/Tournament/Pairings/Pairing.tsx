import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PairingAccordion from './PairingAccordion';
import { RootState } from '../../../app/store';
import type { Match } from './types';
import type { Player } from '../Player/types';
import { alterWithCompletedMatch } from '../Player/utils/player';
import { submitMatchResult } from '../state/tournamentSlice';
import { getExistingMatch } from './utils/match';

interface PairingProps {
  pairing: string[];
  idx: number;
  expandedPairing: number | boolean;
  setExpandedPairing: Function;
}

export const Pairing = (props: PairingProps) => {
  const dispatch = useDispatch();

  const round: number = useSelector(
    (state: RootState) => state.tournament.round
  );
  const players: Player[] = useSelector(
    (state: RootState) => state.tournament.players
  );
  const matchResults: Match[] = useSelector(
    (state: RootState) => state.tournament.matchResults
  );

  const existingMatch = getExistingMatch(props.pairing, matchResults);
  // TODO: error handing for find?
  const firstPlayer: Player = players.find(
    player => player.id === props.pairing[0]
  )!;
  const secondPlayer: Player = players.find(
    player => player.id === props.pairing[1]
  )!;

  return (
    <PairingAccordion
      completedMatch={existingMatch}
      expanded={props.expandedPairing === props.idx}
      handleChange={() => (event: React.SyntheticEvent, isExpanded: boolean) => {
        props.setExpandedPairing(isExpanded ? props.idx : false);
      }}
      firstPlayer={alterWithCompletedMatch(firstPlayer, existingMatch)}
      secondPlayer={alterWithCompletedMatch(secondPlayer, existingMatch)}
      table={props.idx + 1}
    />
  );
};