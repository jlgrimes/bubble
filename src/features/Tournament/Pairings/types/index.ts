import { Match } from "./Match";
import { MatchResult } from "./MatchResult";

export type { Match, MatchResult };