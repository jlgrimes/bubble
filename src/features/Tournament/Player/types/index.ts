import { Player } from "./Player";
import { PlayerMatch } from "./PlayerMatch";
import { Record } from "./Record";

export type { Player, PlayerMatch, Record };