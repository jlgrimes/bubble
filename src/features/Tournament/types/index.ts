import { TournamentState } from "./Tournament";

export type { TournamentState };