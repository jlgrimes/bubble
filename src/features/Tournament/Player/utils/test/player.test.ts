import { Match } from '../../../Pairings/types';
import { Player, PlayerMatch } from '../../../Player/types';
import {
  convertMatchToPlayerMatch,
  getUpdatedPlayerAfterMatch,
  applyMatchResultsToPlayers,
  getUpdatedPlayerPairAfterMatch,
} from '../player';
import {
  SAMPLE_PLAYER,
  SAMPLE_SORTED_PLAYER_LIST,
  SAMPLE_MATCH_RESULTS,
} from '../../../../../helpers/testConstants';

describe('player helpers', () => {
  describe('convertMatchToPlayerMatch', () => {
    it('should preserve match result if player is listed first', () => {
      const match: Match = { result: 'win', playerIds: ['0', '4'] };
      const expectedPlayerMatch: PlayerMatch = {
        opponentId: '4',
        result: 'win',
      };

      expect(convertMatchToPlayerMatch(SAMPLE_PLAYER, match)).toEqual(
        expectedPlayerMatch
      );
    });

    it('should inverse match result if player is listed second', () => {
      const match: Match = { result: 'win', playerIds: ['4', '0'] };
      const expectedPlayerMatch: PlayerMatch = {
        opponentId: '4',
        result: 'loss',
      };

      expect(convertMatchToPlayerMatch(SAMPLE_PLAYER, match)).toEqual(
        expectedPlayerMatch
      );
    });
  });

  describe('getUpdatedPlayerAfterMatch', () => {
    it('should get correct player after match', () => {
      const match: Match = { result: 'win', playerIds: ['0', '4'] };
      const expectedPlayerMatch: PlayerMatch = {
        opponentId: '4',
        result: 'win',
      };
      const expectedPlayer: Player = {
        ...SAMPLE_PLAYER,
        matches: [...SAMPLE_PLAYER.matches, expectedPlayerMatch],
        record: {
          wins: 2,
          ties: 1,
          losses: 1,
        },
        matchPoints: 7,
      };

      expect(getUpdatedPlayerAfterMatch(SAMPLE_PLAYER, match)).toEqual(
        expectedPlayer
      );
    });
  });

  describe('getUpdatedPlayerPairAfterMatch', () => {
    it('should get updated player pair for defined players', () => {
      expect(getUpdatedPlayerPairAfterMatch(SAMPLE_MATCH_RESULTS[0], SAMPLE_SORTED_PLAYER_LIST)).toMatchSnapshot();
    });
  });
  
  describe('applyMatchResultsToPlayers', () => {
    it('should apply match results to players', () => {
      expect(
        applyMatchResultsToPlayers(
          SAMPLE_MATCH_RESULTS,
          SAMPLE_SORTED_PLAYER_LIST
        )
      ).toMatchSnapshot();
    });
  });
});
