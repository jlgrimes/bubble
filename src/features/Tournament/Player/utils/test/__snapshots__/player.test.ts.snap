// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`player helpers applyMatchResultsToPlayers should apply match results to players 1`] = `
Array [
  Object {
    "id": 0,
    "matchPoints": 9,
    "matches": Array [
      Object {
        "opponentId": 3,
        "result": "win",
      },
      Object {
        "opponentId": 1,
        "result": "win",
      },
      Object {
        "opponentId": 4,
        "result": "win",
      },
    ],
    "name": "Jared",
    "record": Object {
      "losses": 0,
      "ties": 0,
      "wins": 3,
    },
  },
  Object {
    "id": 2,
    "matchPoints": 6,
    "matches": Array [
      Object {
        "opponentId": 1,
        "result": "loss",
      },
      Object {
        "opponentId": 3,
        "result": "win",
      },
      Object {
        "opponentId": 5,
        "result": "win",
      },
    ],
    "name": "Ryan",
    "record": Object {
      "losses": 1,
      "ties": 0,
      "wins": 2,
    },
  },
  Object {
    "id": 2,
    "matchPoints": 3,
    "matches": Array [
      Object {
        "opponentId": 1,
        "result": "loss",
      },
      Object {
        "opponentId": 3,
        "result": "win",
      },
      Object {
        "opponentId": 6,
        "result": "loss",
      },
    ],
    "name": "Ryan",
    "record": Object {
      "losses": 2,
      "ties": 0,
      "wins": 1,
    },
  },
  Object {
    "id": 3,
    "matchPoints": 0,
    "matches": Array [
      Object {
        "opponentId": 0,
        "result": "loss",
      },
      Object {
        "opponentId": 0,
        "result": "loss",
      },
      Object {
        "opponentId": 7,
        "result": "loss",
      },
    ],
    "name": "Kenny",
    "record": Object {
      "losses": 3,
      "ties": 0,
      "wins": 0,
    },
  },
]
`;

exports[`player helpers chunkSortedArrayByMatchPoints should chunk array by match points 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    1,
    2,
  ],
  Array [
    3,
  ],
]
`;

exports[`player helpers getUpdatedPlayerPairAfterMatch should get updated player pair for defined players 1`] = `
Array [
  Object {
    "id": 0,
    "matchPoints": 9,
    "matches": Array [
      Object {
        "opponentId": 3,
        "result": "win",
      },
      Object {
        "opponentId": 1,
        "result": "win",
      },
      Object {
        "opponentId": 4,
        "result": "win",
      },
    ],
    "name": "Jared",
    "record": Object {
      "losses": 0,
      "ties": 0,
      "wins": 3,
    },
  },
]
`;
